selection_tools/src/Event.cpp

```diff
@@ -118,15 +118,37 @@ int Event::check_sister(int sister_index, Particle *particle, vector<Particle*>
 {
   // 0: bad particle, -1: opposite sign, +1: same sign
   if(sister_index < 0) return 0;
-  
+  sister_index = get_final_state_particle_index(sister_index, particles);
   auto sister = particles[sister_index];
+
   if(!sister->is_good_non_top_muon(particles)) return 0;
-  
   if(particle->pdgid == sister->pdgid)  return 1;
   else                                  return -1;
 }
 
-int Event::passes_preselection()
+int Event::get_final_state_particle_index(int particle_index, vector<Particle*> particles)
+{
+  bool daughter_exists = true;
+  int new_particle_index = particle_index;
+  while(daughter_exists)
+  {
+    daughter_exists = false;
+    for (auto daughter_index : particles[new_particle_index]->daughters)
+    {
+      if (daughter_index > 0) 
+      {
+        if (particles[daughter_index]->pdgid == particles[particle_index]->pdgid)
+        {
+          daughter_exists = true;
+          new_particle_index = daughter_index;
+        }
+      }
+    }
+  }
+  return new_particle_index;
```

Comment:
Small thing, related to coding conventions - in the rest of the framework I keep the opening curly bracket in the same line as the keyword, except for functions and methods.